File: 49-GroupAnagrams/49-GroupAnagrams.java
// Last updated: 2026. 7. 16. 오전 11:28:35
import java.util.*;

class Solution {
    public String findSortedKey(String str) {
		char[] chars = str.toCharArray();
		Arrays.sort(chars);
		return String.valueOf(chars);
	}

	public List<List<String>> groupAnagrams(String[] strs) {
		Map<String, List<String>> hashMap = new HashMap<>();

		for (String str : strs) {
			hashMap.computeIfAbsent(findSortedKey(str), key -> new ArrayList<>()).add(str);
		}

		return new ArrayList<>(hashMap.values());
	}
}

File: 49-GroupAnagrams/49-GroupAnagrams.cpp
// Last updated: 2026. 7. 16. 오전 11:28:30
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<array<int, 26>, vector<string>> m;
        for (string str : strs) {
            array<int, 26> key = {};
            for (char ch : str) {
                ++key[ch - 'a'];
            }
            m[key].push_back(str);
        }
        vector<vector<string>> result;
        for (auto tuple : m) {
            result.push_back(tuple.second);
        }
        return result;
    }
};